### ISSUE TITLE
Update redshift column encoding from lzo to zstd

### ISSUE THREAD

**Issue by philbooth** [mozilla/fxa-activity-metrics#73]
While investigating #72, I noticed that we're not using the optimum column encoding on our tables any more. Apparently redshift added a new encoding in January, `zstd`. If the claimed reductions are are anything like the numbers below, we should switch to it.

Here's what I found:

```
fxa=# ANALYZE COMPRESSION flow_events;
    Table    |  Column   | Encoding | Est_reduction_pct
-------------+-----------+----------+-------------------
 flow_events | timestamp | zstd     | 58.17
 flow_events | type      | zstd     | 32.36
 flow_events | flow_id   | zstd     | 52.86
 flow_events | flow_time | zstd     | 23.66
 flow_events | locale    | zstd     | 91.98
 flow_events | uid       | zstd     | 91.98
(6 rows)

fxa=# ANALYZE COMPRESSION flow_metadata;
     Table     |    Column    | Encoding | Est_reduction_pct
---------------+--------------+----------+-------------------
 flow_metadata | flow_id      | zstd     | 46.65
 flow_metadata | begin_time   | zstd     | 59.74
 flow_metadata | duration     | zstd     | 19.37
 flow_metadata | completed    | zstd     | 16.32
 flow_metadata | new_account  | zstd     | 43.28
 flow_metadata | ua_browser   | zstd     | 45.54
 flow_metadata | ua_version   | zstd     | 49.29
 flow_metadata | ua_os        | zstd     | 42.58
 flow_metadata | context      | zstd     | 28.67
 flow_metadata | entrypoint   | zstd     | 45.76
 flow_metadata | migration    | zstd     | 88.63
 flow_metadata | service      | zstd     | 48.38
 flow_metadata | utm_campaign | zstd     | 50.26
 flow_metadata | utm_content  | zstd     | 41.69
 flow_metadata | utm_medium   | zstd     | 53.37
 flow_metadata | utm_source   | zstd     | 53.22
 flow_metadata | utm_term     | zstd     | 91.98
 flow_metadata | export_date  | zstd     | 91.54
 flow_metadata | locale       | zstd     | 91.98
 flow_metadata | uid          | zstd     | 91.98
(20 rows)

fxa=# ANALYZE COMPRESSION flow_experiments;
      Table       |   Column    | Encoding | Est_reduction_pct
------------------+-------------+----------+-------------------
 flow_experiments | experiment  | zstd     | 87.10
 flow_experiments | cohort      | zstd     | 87.65
 flow_experiments | timestamp   | zstd     | 49.87
 flow_experiments | flow_id     | zstd     | 46.92
 flow_experiments | uid         | zstd     | 49.59
 flow_experiments | export_date | zstd     | 88.88
(6 rows)

fxa=# ANALYZE COMPRESSION activity_events;
      Table      |   Column   | Encoding | Est_reduction_pct
-----------------+------------+----------+-------------------
 activity_events | timestamp  | zstd     | 61.78
 activity_events | uid        | zstd     | 47.19
 activity_events | type       | zstd     | 22.24
 activity_events | device_id  | zstd     | 44.17
 activity_events | service    | zstd     | 48.38
 activity_events | ua_browser | zstd     | 50.63
 activity_events | ua_version | zstd     | 46.01
 activity_events | ua_os      | zstd     | 44.75
(8 rows)

fxa=# ANALYZE COMPRESSION daily_activity_per_device;
           Table           |   Column   | Encoding | Est_reduction_pct
---------------------------+------------+----------+-------------------
 daily_activity_per_device | day        | zstd     | 91.56
 daily_activity_per_device | uid        | zstd     | 46.74
 daily_activity_per_device | device_id  | zstd     | 44.23
 daily_activity_per_device | service    | zstd     | 51.16
 daily_activity_per_device | ua_browser | zstd     | 75.60
 daily_activity_per_device | ua_version | zstd     | 49.16
 daily_activity_per_device | ua_os      | zstd     | 45.41
(7 rows)

fxa=# ANALYZE COMPRESSION daily_multi_device_users;
          Table           |   Column    | Encoding | Est_reduction_pct
--------------------------+-------------+----------+-------------------
 daily_multi_device_users | day         | zstd     | 91.56
 daily_multi_device_users | uid         | zstd     | 47.01
 daily_multi_device_users | device_now  | zstd     | 44.25
 daily_multi_device_users | device_prev | zstd     | 44.24
(4 rows)
```

/cc @vbudhram, you probably want to start off with `zstd` instead of `lzo` for the email events.

**Comment by philbooth** [comment]
Amazon have a utility than can update this data in-place, which comes with an appropriately scary warning about backing up your data first and so on:

https://github.com/awslabs/amazon-redshift-utils/tree/master/src/ColumnEncodingUtility

We could set up a temporary cluster for a trial run and see if it works out?

**Comment by philbooth** [comment]
> We could set up a temporary cluster for a trial run and see if it works out?

To this end, I created a fresh redshift cluster called `fxa-zstd-test` and am importing the flow event data to it now with `lzo` encoding. That will take a couple of days, after which I'll run the in-place update to `zstd` encoding. I'll record how long it takes, do some basic sanity checking to make sure it hasn't borked anything and we can see what effect it has on data size / query performance.

**Comment by philbooth** [comment]
Trying to run the linked python script against my test cluster, at the moment it's failing with this error:

```
Traceback (most recent call last):
  File "./analyze-schema-compression.py", line 1153, in <module>
    main(sys.argv)
  File "./analyze-schema-compression.py", line 1147, in main
    result_code = run()
  File "./analyze-schema-compression.py", line 860, in run
    report_file_handle = open(report_file, 'w')
TypeError: coercing to Unicode: need string or buffer, bool found
```

Line 860 looks like this:

```
report_file_handle = open(report_file, 'w')
```

I printed the value of `report_file` to the console and it is the correct file path that I'm running the script with. Not sure where this boolean it's complaining about is coming from?

**Comment by hshain** [comment]
same issue -- any help greatly appreciated.

**Comment by rfk** [comment]
If I specify `--report-file REPORT.txt` on the command line then it gets past this error, although I get an authentication error that I haven't yet dug in to.

**Comment by philbooth** [comment]
> If I specify `--report-file REPORT.txt` on the command line then it gets past this error...

Well that's pretty embarrassing, I thought I'd set that option but clearly not. The script now works fine for me, so I've kicked off an in-place update on the spare redshift cluster at 09:57 UTC. I set `\timing` on in the console so I should be able to see how long each of the generated subqueries takes against each table, in addition to how long the whole thing lasts start to finish.

Early indications look pretty good though, for instance it took 3 minutes to completely clone the unsampled `flow_metadata` table.

After it's done I'll run some sanity checks too.

**Comment by philbooth** [comment]
So the whole thing ran in about 20 minutes, but it decided not to update `flow_events` (and its variants) for some reason. Digging in to it at the moment...

**Comment by philbooth** [comment]
There's a `--force` argument to the original script that causes it to emit update code for all the tables. So the full invocation of `analyze-schema-compression.py` (with secrets redacted) now looks like this:

```
./analyze-schema-compression.py --db REDACTED --db-user REDACTED --db-pwd REDACTED --db-host REDACTED --db-port REDACTED --output-file fix-schema-compression.sql --report REPORT.txt --force true
```

Then running the generated update script, again with `\timing` on, looks like this:

```
$ date
Thu  4 May 10:46:26 UTC 2017
$ ./redshift_connect.sh
psql (9.2.18, server 8.0.2)
WARNING: psql version 9.2, server version 8.0.
         Some psql features might not work.
SSL connection (cipher: ECDHE-RSA-AES256-GCM-SHA384, bits: 256)
Type "help" for help.

fxa=# \timing
Timing is on.
fxa=# \i fix-schema-compression.sql
BEGIN
Time: 1.533 ms
LOCK TABLE
Time: 1.042 ms
CREATE TABLE
Time: 166.923 ms
INSERT 0 206147
Time: 1989.456 ms
ANALYZE
Time: 3693.860 ms
ALTER TABLE
Time: 37.670 ms
ALTER TABLE
Time: 1.192 ms
COMMIT
Time: 675.822 ms
BEGIN
Time: 1.322 ms
LOCK TABLE
Time: 0.964 ms
CREATE TABLE
Time: 105.311 ms
INSERT 0 954541
Time: 3187.804 ms
ANALYZE
Time: 233.843 ms
ALTER TABLE
Time: 26.375 ms
ALTER TABLE
Time: 1.174 ms
COMMIT
Time: 471.506 ms
BEGIN
Time: 1.382 ms
LOCK TABLE
Time: 0.960 ms
CREATE TABLE
Time: 192.173 ms
INSERT 0 1872243
Time: 6263.845 ms
ANALYZE
Time: 421.983 ms
ALTER TABLE
Time: 3.523 ms
ALTER TABLE
Time: 1.111 ms
COMMIT
Time: 528.483 ms
BEGIN
Time: 1.300 ms
LOCK TABLE
Time: 1.052 ms
CREATE TABLE
Time: 313.576 ms
INSERT 0 27477340
Time: 49694.619 ms
ANALYZE
Time: 12079.901 ms
ALTER TABLE
Time: 1.219 ms
ALTER TABLE
Time: 1.082 ms
COMMIT
Time: 2773.417 ms
BEGIN
Time: 1.367 ms
LOCK TABLE
Time: 0.961 ms
CREATE TABLE
Time: 6.491 ms
INSERT 0 87490389
Time: 50395.540 ms
ANALYZE
Time: 17563.153 ms
ALTER TABLE
Time: 199.851 ms
ALTER TABLE
Time: 1.206 ms
COMMIT
Time: 4023.559 ms
BEGIN
Time: 1.367 ms
LOCK TABLE
Time: 0.987 ms
CREATE TABLE
Time: 526.459 ms
INSERT 0 104005722
Time: 186142.284 ms
ANALYZE
Time: 61220.399 ms
ALTER TABLE
Time: 1.311 ms
ALTER TABLE
Time: 1.209 ms
COMMIT
Time: 706.552 ms
BEGIN
Time: 1.497 ms
LOCK TABLE
Time: 1.064 ms
CREATE TABLE
Time: 157.660 ms
INSERT 0 101902061
Time: 230391.631 ms
ANALYZE
Time: 56421.204 ms
ALTER TABLE
Time: 1.321 ms
ALTER TABLE
Time: 1.224 ms
COMMIT
Time: 522.155 ms
BEGIN
Time: 1.407 ms
LOCK TABLE
Time: 1.031 ms
CREATE TABLE
Time: 5.692 ms
INSERT 0 371143559
Time: 247957.675 ms
ANALYZE
Time: 69091.092 ms
ALTER TABLE
Time: 1.409 ms
ALTER TABLE
Time: 1.233 ms
COMMIT
Time: 280.776 ms
BEGIN
Time: 1.326 ms
LOCK TABLE
Time: 0.967 ms
CREATE TABLE
Time: 5.952 ms
INSERT 0 490686133
Time: 330502.083 ms
ANALYZE
Time: 95930.299 ms
ALTER TABLE
Time: 1.256 ms
ALTER TABLE
Time: 1.109 ms
COMMIT
Time: 372.498 ms
fxa=# \q
$ date
Thu  4 May 11:10:51 UTC 2017
```

So ~25 minutes in total with the longest individual query taking 5 and a half minutes.

Obviously, this doesn't include the activity event data but that still seems pretty fast to me.

The only thing that suprised me was, even though the report suggests the best encoding for `timestamp`/`begin_time` is `zstd`, the generated SQL encodes them all as `RAW`. However, the following docs suggest that is the correct call, presumably to improve query performance:

http://docs.aws.amazon.com/redshift/latest/dg/c_Raw_encoding.html

I compared the data for a few different days between the updated cluster and our real cluster, everything appears to match. And the queries are definitely running significantly quicker on the updated cluster, but that may be just because they don't have a million scheduled queries hogging all the resources.

Bottom line, I'm feeling quite gung-ho about this now. Is there anything else I should be checking before we do this for real? Can I have a bash at doing it for real this weekend? 😄

**Comment by philbooth** [comment]
> Is there anything else I should be checking before we do this for real?

I'll need to prepare a PR to update the import scripts too of course. I'll do that presently.

**Comment by philbooth** [comment]
(btw, I haven't terminated the zstd cluster yet in case anyone wants to go in and run some checks themselves, but if people are happy I can kill it to stop accruing further charges)

**Comment by philbooth** [comment]
> I'll need to prepare a PR to update the import scripts too of course. I'll do that presently.

#78

**Comment by rfk** [comment]
> Bottom line, I'm feeling quite gung-ho about this now. Is there anything else I should be
> checking before we do this for real? Can I have a bash at doing it for real this weekend?

Shipit!

**Comment by philbooth** [comment]
Of course, running smoothly on the test cluster is no guarantee of anything and, almost predictably, the analysis script is repeatedly failing on `rfkelly-fxa-test`.

```
-- [16282] Analyzing Table 'flow_metadata'
(u'ERROR', u'XX000', u'abort query', u'\n  -----------------------------------------------\n  error:  abort query\n  code:      1020\n  context:   system requested abort\n  query:     14544\n  location:  queryabort.cpp:423\n  process:   padbmaster [pid=19067]\n  -----------------------------------------------\n', u'/home/ec2-user/padb/src/sys/xen_execute.cpp', u'6589', u'pg_throw', u'', u'')
(u'ERROR', u'25P02', u'current transaction is aborted, commands ignored until end of transaction block', u'/home/ec2-user/padb/src/pg/src/backend/tcop/postgres.c', u'2922', u'exec_bind_message', u'', u'')
(u'ERROR', u'25P02', u'current transaction is aborted, commands ignored until end of transaction block', u'/home/ec2-user/padb/src/pg/src/backend/tcop/postgres.c', u'2922', u'exec_bind_message', u'', u'')
(u'ERROR', u'25P02', u'current transaction is aborted, commands ignored until end of transaction block', u'/home/ec2-user/padb/src/pg/src/backend/tcop/postgres.c', u'2922', u'exec_bind_message', u'', u'')
(u'ERROR', u'25P02', u'current transaction is aborted, commands ignored until end of transaction block', u'/home/ec2-user/padb/src/pg/src/backend/tcop/postgres.c', u'2922', u'exec_bind_message', u'', u'')
(u'ERROR', u'25P02', u'current transaction is aborted, commands ignored until end of transaction block', u'/home/ec2-user/padb/src/pg/src/backend/tcop/postgres.c', u'2922', u'exec_bind_message', u'', u'')
(u'ERROR', u'25P02', u'current transaction is aborted, commands ignored until end of transaction block', u'/home/ec2-user/padb/src/pg/src/backend/tcop/postgres.c', u'2922', u'exec_bind_message', u'', u'')
(u'ERROR', u'25P02', u'current transaction is aborted, commands ignored until end of transaction block', u'/home/ec2-user/padb/src/pg/src/backend/tcop/postgres.c', u'2922', u'exec_bind_message', u'', u'')
(u'ERROR', u'25P02', u'current transaction is aborted, commands ignored until end of transaction block', u'/home/ec2-user/padb/src/pg/src/backend/tcop/postgres.c', u'2922', u'exec_bind_message', u'', u'')
(u'ERROR', u'25P02', u'current transaction is aborted, commands ignored until end of transaction block', u'/home/ec2-user/padb/src/pg/src/backend/tcop/postgres.c', u'2922', u'exec_bind_message', u'', u'')
Unable to analyze flow_metadata due to Exception
Traceback (most recent call last):
  File "./analyze-schema-compression.py", line 1154, in <module>
    main(sys.argv)
  File "./analyze-schema-compression.py", line 1148, in main
    result_code = run()
  File "./analyze-schema-compression.py", line 958, in run
    result.append(analyze(t))
  File "./analyze-schema-compression.py", line 398, in analyze
    output = get_count_raw_columns(table_name)
  File "./analyze-schema-compression.py", line 361, in get_count_raw_columns
    description = execute_query(statement)
  File "./analyze-schema-compression.py", line 90, in execute_query
    cursor.execute(str)
  File "/usr/lib/python2.7/site-packages/pg8000/core.py", line 906, in execute
    self._c.execute(self, operation, args)
  File "/usr/lib/python2.7/site-packages/pg8000/core.py", line 1940, in execute
    self.handle_messages(cursor)
  File "/usr/lib/python2.7/site-packages/pg8000/core.py", line 2088, in handle_messages
    raise self.error
pg8000.core.ProgrammingError: (u'ERROR', u'25P02', u'current transaction is aborted, commands ignored until end of tra  File "./analyze-schema-compression.py", line 90, in execute_query
    cursor.execute(str)
  File "/usr/lib/python2.7/site-packages/pg8000/core.py", line 906, in execute
    self._c.execute(self, operation, args)
  File "/usr/lib/python2.7/site-packages/pg8000/core.py", line 1940, in execute
    self.handle_messages(cursor)
  File "/usr/lib/python2.7/site-packages/pg8000/core.py", line 2088, in handle_messages
    raise self.error
pg8000.core.ProgrammingError: (u'ERROR', u'25P02', u'current transaction is aborted, commands ignored until end of transaction block', u'/home/ec2-user/padb/src/pg/src/backend/tcop/postgres.c', u'2441', u'exec_parse_message', u'', u'')
```

I've tried a few times, with and without `--force`, with different numbers for `--threads`, but it always fails with that error.

So what I'll probably do instead, tomorrow when I have more time, is modify the generated script from the test cluster so it includes the activity event tables, then run that instead.

**Comment by philbooth** [comment]
This just finished now. It took about 13 hours in the end, mostly because there were long waits to lock tables with all the scheduled queries running.

The pre-migration tables are only renamed, not dropped, so we can still switch back very easily if we want to. I'm doing some sanity checks right now. And if anyone else spots anything wrong just let me know.

**Comment by philbooth** [comment]
I am officially calling this a resounding success. [My monster query for phase 2 of connect another device](https://sql.telemetry.mozilla.org/queries/4198/source#8355) just completed in 18 minutes. Previous best was 2 and a half hours. Closing.

**Comment by philbooth** [comment]
...and I'll delete the old tables at the end of today, unless anyone tells me not to.
